File: 2017-javascript/day01.js
import { importInput } from "./helpers.js";
const input = importInput("./inputs/1");
function findDuplicates(str = "") {
  let arr = str
    .trim()
    .split("")
    .filter((el, index, arr) => {
      if (index === arr.length - 1) {
        return el === arr[0];
      } else {
        return el === arr[index + 1];
      }
    });
  return arr.reduce((a, b) => +a + +b, 0);
}

function findOffsetDuplicates(str = "") {
  let arr = str
    .trim()
    .split("")
    .filter((el, index, arr) => {
      let offset = arr.length / 2;
      if (index + offset >= arr.length) {
        return el === arr[offset - (arr.length - index)];
      } else {
        return el === arr[index + offset];
      }
    });
  return arr.reduce((a, b) => +a + +b, 0);
}

let partOne = findDuplicates(input);
let partTwo = findOffsetDuplicates(input);
console.log("===== Day 1 =====");
console.log("Part One: ", partOne);
console.log("Part Two: ", partTwo);

File: 2020-python/day01.py
input = open('./inputs/1', 'r').read().split('\n')
input.pop()
# part one
for i in range(len(input)):
    for j in range(i, len(input)):
        if int(input[i]) + int(input[j]) == 2020:
            print(int(input[i]) * int(input[j]))

# part two
# optimal 3-sum
input = sorted(list(map(lambda x: int(x), input)))

for i in range(len(input)-2):
    if i > 0 and input[i] == input[i-1]:
        continue
    l = i + 1
    r = len(input) - 1
    while l < r:
        if input[i] + input[l] + input[r] == 2020:
            print(input[i] * input[l] * input[r])
            break
        elif input[i] + input[l] + input[r] < 2020:
            l += 1
        else:
            r -= 1

